src/persist-client/src/internal/state_diff.rs

```diff
@@ -0,0 +1,576 @@
+                    "{} update didn't match: {:?} vs {:?}",
+                    name, single, &from
+                ));
+            }
+            assert_eq!(single, &from);
+            *single = to
+        }
+        Insert(_) => return Err(format!("cannot insert {} field", name)),
+        Delete(_) => return Err(format!("cannot delete {} field", name)),
+    }
+    Ok(())
+}
+
+fn diff_field_sorted_iter<'a, K, V, IF, IT>(from: IF, to: IT, diffs: &mut Vec<StateFieldDiff<K, V>>)
+where
+    K: Ord + Clone + 'a,
+    V: PartialEq + Clone + 'a,
+    IF: IntoIterator<Item = (&'a K, &'a V)>,
+    IT: IntoIterator<Item = (&'a K, &'a V)>,
+{
+    let (mut from, mut to) = (from.into_iter(), to.into_iter());
+    let (mut f, mut t) = (from.next(), to.next());
+    loop {
+        match (f, t) {
+            (None, None) => break,
+            (Some((fk, fv)), Some((tk, tv))) => match fk.cmp(tk) {
+                Ordering::Less => {
+                    diffs.push(StateFieldDiff {
+                        key: fk.clone(),
+                        val: Delete(fv.clone()),
+                    });
+                    f = from.next();
+                }
+                Ordering::Greater => {
+                    diffs.push(StateFieldDiff {
+                        key: tk.clone(),
+                        val: Insert(tv.clone()),
+                    });
+                    t = to.next();
+                }
+                Ordering::Equal => {
+                    // WIP regression test for this if, I missed it in the
+                    // original impl :)
+                    if fv != tv {
+                        diffs.push(StateFieldDiff {
+                            key: fk.clone(),
+                            val: Update(fv.clone(), tv.clone()),
+                        });
+                    }
+                    f = from.next();
+                    t = to.next();
+                }
+            },
+            (None, Some((tk, tv))) => {
+                diffs.push(StateFieldDiff {
+                    key: tk.clone(),
+                    val: Insert(tv.clone()),
+                });
+                t = to.next();
+            }
+            (Some((fk, fv)), None) => {
+                diffs.push(StateFieldDiff {
+                    key: fk.clone(),
+                    val: Delete(fv.clone()),
+                });
+                f = from.next();
+            }
+        }
+    }
+}
+
+fn apply_diffs_map<K: Ord, V: PartialEq + Debug>(
+    name: &str,
+    diffs: Vec<StateFieldDiff<K, V>>,
+    map: &mut BTreeMap<K, V>,
+) -> Result<(), String> {
+    for diff in diffs {
+        apply_diff_map(name, diff, map)?;
+    }
+    Ok(())
+}
+
+fn apply_diff_map<K: Ord, V: PartialEq + Debug>(
+    name: &str,
+    diff: StateFieldDiff<K, V>,
+    map: &mut BTreeMap<K, V>,
+) -> Result<(), String> {
+    match diff.val {
+        Insert(to) => {
+            let prev = map.insert(diff.key, to);
+            if prev != None {
```

Comment:
All these mutate our state and then return an error if the application was incorrect. Maybe these should check before applying the diff, so as not to leave the state in an incorrect (um) state? As it is, we rely on the caller cleaning up or panic!'ing. Could also be overly cautious, though.

Reply:
yeah, I thought about this, but it makes the code soooo much harder to write. given that we always panic on error, I lean toward punt. i'll drop a note here though, so future readers will know we considered this